fix(appmesh): introduce the TlsClientPolicy and TlsValidation concepts

Seiya6329 opened the pull request:
This is the first part of the series for [issue#12733](#12733).
For the breakdown on series, please refer to
[this comment](#14782 (comment)).

#### Collaborators
@alexbrjo and @dfezzie. Thank you!

#### REV:
- Adding `TlsValidation` abstract class with factory methods for available certificate sources.
- Converting client policy into an interface. In addition, it is renamed to `TlsClientPolicy` to better reflect its representation.

#### Design Note:
- Adding `TlsValidation` abstract class since it became a common property for TLS client policy and TLS listener after mTLS feature support.

BREAKING CHANGE: the creation property `clientPolicy` in `VirtualNode` has been renamed to `tlsClientPolicy`, and its type changed to `TlsClientPolicy`
- **appmesh**: the creation property `clientPolicy` in `VirtualGateway` has been renamed to `tlsClientPolicy`, and its type changed to `TlsClientPolicy`
- **appmesh**: to create `TlsClientPolicy`, `validation` property must be defined.

---
*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 9947cfda3b19472067d4940baa4ab11560b9d61a
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d8db51b7-b241-4719-b7c3-98b447d01ee6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:
```diff
@@ -167,6 +167,31 @@ class FileAccessLog extends AccessLog {
+   * TLS is enforced on the ports specified here.
+   * If no ports are specified, TLS will be enforced on all the ports.
+   *
+   * @default: none
```
1. No need for the `:` here.
2. The description of the default is not consistent with the description of the property. You probably want something like:

```suggestion
   * @default - TLS will be enforced on all ports
```

Seiya6329 replied:
Great point! I will simply this a bit and say "all ports".

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:
```diff
@@ -238,7 +263,17 @@ class VirtualServiceBackend extends Backend {
+                ports: this.tlsClientPolicy.ports,
+                enforce: this.tlsClientPolicy.enforce,
+                validation: {
+                  trust: this.tlsClientPolicy.tlsValidationContext.trust.bind(_scope).virtualNodeClientTlsValidationContextTrust,
```
Now that it's being used, `_scope` should be renamed to `scope`.

Seiya6329 replied:
Addressed.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:
```diff
@@ -167,6 +167,31 @@ class FileAccessLog extends AccessLog {
+  /**
+   * Represents the object for TLS validation context
+   */
+  readonly tlsValidationContext: TlsValidationContext;
```
I don't like this property. This is needless nesting that doesn't add anything to the API. Remove this, and instead add a `readonly trust: TlsValidationContextTrust;` property.

Seiya6329 replied:
This was decided to isolate the works in steps. To be little more in detail:
**Step1:** Create TLS validation object since it is going to be used by TLS Listener in addition to TLS client policy (*I am here)
**Step2:** Update TLS certificate object since it contains `mode` property that is actually the property for TLS Listener. In addition, TLS certificate is going to be used by TLS client policy.
**Step3:** Implement mTLS feature. `subjectiveAlternativeNames` addition is specific to mTLS feature.

skinny85 replied:
Ah. So what you're saying is that `TlsValidationContext` will be used in 2 separate places - in `TlsClientPolicy`, and then somewhere else?

Seiya6329 replied:
Yes, that is exactly correct. Please refer [AWS::AppMesh::VirtualNode ListenerTls](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-virtualnode-listenertls.html) and [AWS::AppMesh::VirtualNode ClientPolicyTls](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-virtualnode-clientpolicytls.html) as an example. As you may see, both objects have properties `certificate` and `validation`.

skinny85 replied:
OK. You've convinced me 🙂. Let's keep this interface, but rename it to `TlsValidation` (I don't think the word "context" adds much value here).

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:
```diff
@@ -167,6 +167,31 @@ class FileAccessLog extends AccessLog {
+/**
+ * Represents the properties needed to define client policy
+ */
+export interface TlsClientPolicy {
```
Let's move this interface to the same file that defines `TlsValidationContext`, and the remaining classes/interfaces there.

Seiya6329 replied:
Could you advise the intention for moving `TlsClientPolicy` interface? If this is happening, I am wondering if `BackendDefaults` should also move too,

```
export interface BackendDefaults {
  /**
   * Client policy for backend defaults
   *
   * @default - none
   */
  readonly tlsClientPolicy?: TlsClientPolicy;
}
```

skinny85 replied:
> Could you advise the intention for moving TlsClientPolicy interface?

I don't like files with generic names like `shared-interfaces`. Over time, the become dumping grounds for things, and they get very long. Things should be grouped logically - things are logically close together, like `TlsClientPolicy` and `TlsValidation`, should be in one place.

> If this is happening, I am wondering if BackendDefaults should also move too/

Probably, but this is unrelated to your change, so feel free to keep them here for this PR.

Seiya6329 replied:
Thanks Adam! It makes sense that overall motivation is to avoid using generic file. Will move `TlsClientPolicy` interface as advised.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:
```diff
@@ -238,7 +263,17 @@ class VirtualServiceBackend extends Backend {
         virtualService: {
           virtualServiceName: this.virtualService.virtualServiceName,
-          clientPolicy: this.clientPolicy?.bind(_scope).clientPolicy,
+          clientPolicy: this.tlsClientPolicy !== undefined ?
+            {
```
Indentation should be different here:

```suggestion
          clientPolicy: this.tlsClientPolicy !== undefined
            ? {
```

Seiya6329 replied:
Addressed

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation-context.ts:
```diff
@@ -0,0 +1,160 @@
+// eslint-disable-next-line no-duplicate-imports, import/order
+import { Construct } from '@aws-cdk/core';
+
+/**
+ * Represents the properties needed to define TLS validation context
+ */
+export interface TlsValidationContext {
+  /**
+   * Reference to where to retrieve the trust chain.
+   */
+  readonly trust: TlsValidationContextTrust;
+}
```
This interface is redundant. I would remove it.

Seiya6329 replied:
Please refer [this response](https://github.com/aws/aws-cdk/pull/14782#discussion_r636273194) and advise if it makes to retain `TlsValidationContext` interface.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation-context.ts:
```diff
@@ -0,0 +1,160 @@
+  /**
+   * VirtualNode CFN configuration for client policy's TLS Validation
+   */
+  readonly virtualNodeClientTlsValidationContextTrust: CfnVirtualNode.TlsValidationContextTrustProperty;
```
I would remove "context" from the name of all of these properties:

```suggestion
  readonly virtualNodeClientTlsValidationTrust: CfnVirtualNode.TlsValidationContextTrustProperty;
```

Seiya6329 replied:
addressed.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation-context.ts:
```diff
@@ -0,0 +1,160 @@
+   *
+   * @default - no TLS Validation
+   */
+  readonly virtualNodeListenerTlsValidationContextTrust?: CfnVirtualNode.ListenerTlsValidationContextTrustProperty
```
```suggestion
  readonly virtualNodeListenerTlsValidationTrust?: CfnVirtualNode.ListenerTlsValidationContextTrustProperty
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation-context.ts:
```diff
@@ -0,0 +1,160 @@
+  /**
+   * VirtualGateway CFN configuration for client policy's TLS Validation
+   */
+  readonly virtualGatewayClientTlsValidationContextTrust: CfnVirtualGateway.VirtualGatewayTlsValidationContextTrustProperty;
```
```suggestion
  readonly virtualGatewayClientTlsValidationTrust: CfnVirtualGateway.VirtualGatewayTlsValidationContextTrustProperty;
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation-context.ts:
```diff
@@ -0,0 +1,160 @@
+   *
+   * @default - no TLS Validation
+   */
+  readonly virtualGatewayListenerTlsValidationContextTrust?: CfnVirtualGateway.VirtualGatewayListenerTlsValidationContextTrustProperty;
```
```suggestion
  readonly virtualGatewayListenerTlsValidationTrust?: CfnVirtualGateway.VirtualGatewayListenerTlsValidationContextTrustProperty;
```

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation-context.ts:
```diff
@@ -0,0 +1,160 @@
+/**
+ * Defines the TLS validation context trust.
+ */
+export abstract class TlsValidationContextTrust {
```
I don't see what the word "context" adds here. I would rename this to `TlsValidationTrust`:

```suggestion
export abstract class TlsValidationTrust {
```

Seiya6329 replied:
Addressed.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation-context.ts:
```diff
@@ -0,0 +1,160 @@
+  /**
+   * Tells envoy where to fetch the validation context from
+   */
+  public static fileTrust(props: FileTrustOptions): TlsValidationContextTrust {
```
Now that "trust" is in the name, I would rename this to just `file`:

```suggestion
  public static file(props: TlsValidationFileTrustOptions): TlsValidationTrust {
```

Seiya6329 replied:
Good point! Addressed.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation-context.ts:
```diff
@@ -0,0 +1,160 @@
+/**
+ * ACM Trust Properties
+ */
+export interface AcmTrustOptions {
```
This needs to be renamed to `TlsValidationAcmTrustOptions`:

```suggestion
export interface TlsValidationAcmTrustOptions {
```

Seiya6329 replied:
Addressed.

skinny85 reviewed (changes requested):
Looks great @Seiya6329, thanks for the contribution! A few small comments before we can merge this in 🙂.

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:
```diff
@@ -167,6 +167,31 @@ class FileAccessLog extends AccessLog {
+  /**
+   * Whether the policy is enforced.
+   *
+   * @default: true
```
Removed `:` and changed to `@default - ` to match pattern.

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -173,15 +173,26 @@ export class VirtualGateway extends VirtualGatewayBase {
+                  },
+                },
+              }
+              : undefined,
```
I think you are asking to create method/function to assign `tls` properties. For that case, would you recommend to:

- Create `bind()` class for `TlsClientPolicy` (*this will also require to create a class)
- Create `render()` class for each classes that assign `tls` properties.

skinny85 replied:
I would suggest creating a helper function that takes in a `TlsClientPolicy`, and a function that extracts a `trust` property from the argument, something like this:

```ts
function renderTlsPolicy(scope: Construct, tlsClientPolicy: TlsClientPolicy | undefined, extractor: (c: TlsValidationConfig) => any): any {
  return tlsClientPolicy !== undefined
    ?  {
      tls: { 
        ports: tlsClientPolicy.ports, 
        enforce: tlsClientPolicy.enforce, 
        validation: {
          trust: extractor(tlsClientPolicy.trust.bind(scope))
        },
      },
    }
   : undefined,
}
```

Something like this (the types are not great here, but I think the duplication might be worse).

Seiya6329 replied:
Well noted. I will create `render()` helper function for each classes.

skinny85 commented:
BTW - the PR also lacks updating the ReadMe file, and needs to correctly list all breaking changes that happened in this PR - see here: https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md#step-4-commit on how we show breaking changes in the description of the PR.

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation-context.ts:
```diff
@@ -0,0 +1,160 @@
+/**
+ * File Trust Properties
+ */
+export interface FileTrustOptions {
```
Addressed.

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation-context.ts:
```diff
@@ -0,0 +1,160 @@
+  /**
+   * TLS validation context trust for ACM Private Certificate Authority (CA).
+   */
+  public static acmTrust(props: AcmTrustOptions): TlsValidationContextTrust {
```
addressed

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-node.ts:
```diff
@@ -382,10 +386,14 @@ export = {
-          }),
+            tlsValidationContext: {
+              trust: appmesh.TlsValidationContextTrust.fileTrust({
+                certificateChain: 'path/to/cert',
```
Addressed.

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/test/test.virtual-gateway.ts:
```diff
@@ -390,9 +390,13 @@ export = {
+          tlsClientPolicy: {
+            tlsValidationContext: {
+              trust: appmesh.TlsValidationContextTrust.fileTrust({
+                certificateChain: 'path/to/cert',
```
Addressed.

skinny85 replied:
Nope, it is not 🙂.

Seiya6329 replied:
Sorry for the miss. It is now really taken care of!

Seiya6329 commented:
### Note:
I am submitting new commits instead of performing `git commit --amend --no-edit`. When I `--amend`, I was required to force push and the history on the pull request seemed to be broken.

### REV2:
- Addressing the suggestions for renaming.
- Addressing the suggestion to move the location for interfaces.
- Adding `renderTlsPolicy()` method to parse `tls` property.
- Updated `README.md` file

### Design Note:
- Retaining `TlsValidation` interface since we have a plan to add more property to it.

### BREAKING CHANGE:  all modules that uses client policy as a property
* **virtual-gateway:** major update on `ClientPolicy` object causes to break
* **virtual-node:** major update on `ClientPolicy` object causes to break

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: e94fcec19b4e110673cb0b86f22ae46a82f3de83
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=973280dc-627d-4ac2-be66-2bc0d7463ed8%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: f8b5e6c4be674e6cb49dc71bda5f8999542d2205
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=da99ec5a-39c0-4f5a-b25e-b5af81c71c84%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts:
```diff
@@ -0,0 +1,196 @@
+  }
+}
+
+export function renderTlsPolicy(scope: Construct, tlsClientPolicy: TlsClientPolicy,
```
Why is this exported from this file? This makes it part of the public API of the AppMesh module, which I'm pretty sure we don't want.

If you need this in multiple files, then you need to move this function to a separate file, preferably in the `lib/private` subdirectory, so it's obvious it's part of the public API. There's a file called `utils.ts` there, maybe this is a good place for this function.

Seiya6329 replied:
You are right that we do not want to make this function public. I was not sure where to put and I will follow your instruction. Thank you!

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -173,15 +174,19 @@ export class VirtualGateway extends VirtualGatewayBase {
       spec: {
         listeners: this.listeners.map(listener => listener.listener),
-        backendDefaults: props.backendDefaults !== undefined
+        backendDefaults: props.backendDefaults
```
I don't think there's any need for this change, I would leave it as-is.

Seiya6329 replied:
I am ok leaving `! == undefined` but wouldn't this be simply redundant?

skinny85 replied:
It is redundant, but that just means there's little point in changing it. Generally, the smaller the PR, the better 🙂. That's the only motivation I have here.

Seiya6329 replied:
Makes sense! I agree `!== undefined` is not a necessary change and thus leaving it as-is to keep the PR smaller.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -173,15 +174,19 @@ export class VirtualGateway extends VirtualGatewayBase {
           ? {
-            clientPolicy: props.backendDefaults?.clientPolicy?.bind(this).clientPolicy,
+            clientPolicy: {
+              tls: tlsClientPolicy ? renderTlsPolicy(this, tlsClientPolicy,
+                (config) => config.virtualGatewayClientTlsValidationTrust) : undefined,
```
Let's make `renderTlsPolicy` handle the undefined case, we'll be able to say here then:

```suggestion
              tls: renderTlsPolicy(this, tlsClientPolicy, (config) => config.virtualGatewayClientTlsValidationTrust),
```

Much simpler.

Seiya6329 replied:
I tried this path and realized there is a tradeoff to make. 
Currently, `renderTlsPolicy` returns type = `CfnVirtualNode.ClientPolicyTlsProperty`. This type does not allow `undefined` to be assigned. 
In order to make `renderTlsPolicy` handle the undefined case, it must return type = any. I tried to avoid using `any` type since you mentioned [here](https://github.com/aws/aws-cdk/pull/14782#discussion_r636325552) types did not look good.

skinny85 replied:
Simple solution: make `renderTlsPolicy()` return `CfnVirtualNode.ClientPolicyTlsProperty | undefined`. Problem solved 🙂.

Seiya6329 replied:
Of course! I somehow overlooked that union type can be used for the return type as well.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -1,13 +1,19 @@
 import * as cdk from '@aws-cdk/core';
+
```
Why this empty line here suddenly...?

Seiya6329 replied:
This is my miss.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -1,13 +1,19 @@
+
+// keep this import separate from other imports to reduce chance for merge conflicts with v2-main
+// eslint-disable-next-line no-duplicate-imports, import/order
+import { Construct as CoreConstruct } from '@aws-cdk/core'; all VirtualNode based classes must implement
```
Not sure what happened here? Why this comment block here, with the import of Construct?

Seiya6329 replied:
Great catch! Originally, I attempted to add `render TlsPolicy` function to this class and in order to make it work I needed to import `Construct` from `@aws-cdk/core`. I thought I deleted the import statement but I must have missed it.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -176,16 +182,20 @@ export class VirtualNode extends VirtualNodeBase {
+            },
           }
           : undefined,
         serviceDiscovery: {
```
Same comments here (leave the previous condition unchanged, making `renderTlsPolicy()` handle the `undefined` case simplifies this code).

Seiya6329 replied:
Responded [here](https://github.com/aws/aws-cdk/pull/14782#discussion_r637160804).

Seiya6329 replied:
Addressing this as well.

skinny85 reviewed (dismissed):
Looks great @Seiya6329! We're getting very close to having this merged 🙂.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 3937a585df7ef7c84e5a8fa3e341293dff208a13
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8229b31c-7bb5-44aa-806f-58655aa7f892%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -183,10 +183,14 @@ const node = new VirtualNode(this, 'node', {
-      certificateChain: '/keys/local_cert_chain.pem',
-    }),
-  },
+      tlsClientPolicy: {
+        tlsValidationContext: {
+          trust: appmesh.TlsValidationTrust.file({
+            certificateChain: '/keys/local_cert_chain.pem',
+          }),
+        },
+      },
+    },
```
Thanks for the pointer!

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/shared-interfaces.ts:
```diff
@@ -226,19 +226,24 @@ export abstract class Backend {
-          clientPolicy: this.clientPolicy?.bind(_scope).clientPolicy,
+          clientPolicy: this.tlsClientPolicy
+            ? {
+              tls: this.tlsClientPolicy ? renderTlsPolicy(scope, this.tlsClientPolicy,
+                (config) => config.virtualNodeClientTlsValidationTrust) : undefined,
```
You are absolutely right. Will address this for all locations that uses `renderTlspolicy()`.

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -218,12 +222,8 @@ const node = new VirtualNode(this, 'node', {
-    ports: [8080, 8081],
-  }),
+  virtualServiceProvider: appmesh.VirtualServiceProvider.virtualRouter(router),
+  virtualServiceName: 'service1.domain.local',
```
Thank you!

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -493,11 +493,15 @@ const gateway = new appmesh.VirtualGateway(stack, 'gateway', {
+          }),
+        },
+      },
+    },
```
Addressed.

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts:
```diff
@@ -0,0 +1,196 @@
+  /**
+   * Represents the object for TLS validation context
+   */
+  readonly tlsValidationContext: TlsValidation;
```
It makes sense since this is under `TlsClientPolicy`

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts:
```diff
@@ -0,0 +1,196 @@
+
+}
+
+class TlsValidationAcmTrustImpl extends TlsValidationTrust {
```
Addressed.

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts:
```diff
@@ -0,0 +1,196 @@
+  }
+}
+
+class TlsValidationFileTrustImpl extends TlsValidationTrust {
```
Addressed.

Seiya6329 commented:
### REV3:
- Moving `renderTlsPolicy` method to private file. Also renamed it to `renderTlsClientPolicy`.
- Fixed `README.md` file with incorrect indentation.

### Note:
- I have not updated `renderTlsClientPolicy` to handle the undefined case in this change. Please refer my response with the tradeoffs and advise the thought.

### Test:
`yarn run build` and `yarn run test` successful.

### BREAKING CHANGE: 
* **virtual-gateway:** major update on `ClientPolicy` object causes to break
* **virtual-node:** major update on `ClientPolicy` object causes to break

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 1c6cf6f2ea7e0adcfc1f5cedb2642e5d51c1378b
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=20ec7f88-9b50-430a-ad89-3355bb9a68ce%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

Seiya6329 commented:
### REV3:
- Moving `renderTlsPolicy` method to private file. Also renamed it to `renderTlsClientPolicy`.
- Fixed `README.md` file with incorrect indentation.
- Updating `renderTlsClientPolicy` to handle the undefined case

### Test:
`yarn run build` and `yarn run test` successful.

### BREAKING CHANGE: 
* **virtual-gateway:** major update on `ClientPolicy` object causes to break
* **virtual-node:** major update on `ClientPolicy` object causes to break

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 798ef7e9a1ea517e1132fe00b68c3fe0287901d1
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=15dfa2ef-1bea-46f3-874e-1a4510fe6e03%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -183,9 +183,13 @@ const node = new VirtualNode(this, 'node', {
-      certificateChain: '/keys/local_cert_chain.pem',
-    }),
+    tlsClientPolicy: {
+      tlsValidationContext: {
```
This is now called `validation`:

```suggestion
      validation: {
```

Seiya6329 replied:
Thanks for the catch!

skinny85 reviewed (changes requested):
Looks almost ready! I just have a few closing comments/questions.

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -493,10 +493,14 @@ const gateway = new appmesh.VirtualGateway(stack, 'gateway', {
+    tlsClientPolicy: {
       ports: [8080, 8081],
-    }),
+      tlsValidationContext: {
```
Addressed.

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -173,6 +174,7 @@ export class VirtualGateway extends VirtualGatewayBase {
     }
 
     const accessLogging = props.accessLog?.bind(this);
+    const tlsClientPolicy = props.backendDefaults?.tlsClientPolicy;
```
Got it. Thanks for the providing the detail on this change.

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -176,6 +177,7 @@ export class VirtualNode extends VirtualNodeBase {
     props.listeners?.forEach(listener => this.addListener(listener));
     const accessLogging = props.accessLog?.bind(this);
     const serviceDiscovery = props.serviceDiscovery?.bind(this);
+    const tlsClientPolicy = props.backendDefaults?.tlsClientPolicy;
```
Addressed.

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts:
```diff
@@ -0,0 +1,185 @@
+      throw new Error('you must provide at least one Certificate Authority when creating an ACM Trust ClientPolicy');
+    } else {
+      return {
+        virtualNodeClientTlsValidationTrust: {
+          acm: {
+            certificateAuthorityArns: this.certificateAuthorities.map(certificateArn =>
+              certificateArn.certificateAuthorityArn),
+          },
+        },
+        virtualGatewayClientTlsValidationTrust: {
+          acm: {
+            certificateAuthorityArns: this.certificateAuthorities.map(certificateArn =>
+              certificateArn.certificateAuthorityArn),
+          },
+        },
```
This is great question.

> Can you help me understand one thing? Why do all 4 of the L1 TlsValidationContextTrustProperty interfaces have the file property, but only 2 of them have the acm property?

This is due to [mTLS feature limitation](https://docs.aws.amazon.com/app-mesh/latest/userguide/mutual-tls.html) that mTLS only supports `file` and `sds` (*this will be added during my subsequent code change) but NOT `acm`.

> Why have 4 properties, if you only ever use 2 of them?

This is very fair question and I think this confusion is brought up since I included the change to add `virtualNodeListenerTlsValidationTrust` and`virtualGatewayListenerTlsValidationTrust`.  These two properties are specific to mTLS feature implementation and **I shouldn't have included them** in this code review. Will remove them.

skinny85 replied:
Hmm. So how will you model this in the CDK, so that places that don't allow passing `acm` in the service do not allow passing `TlsValidationTrust.acm()` in the CDK?

Ideally, this would be a mistake that we prevent at compile-time.

Seiya6329 replied:
That is another great question. We had a team discussion on whether to enforce the compile-time error here and we concluded not to.
Our reasoning was that `acm` could be supported in future and we would like to avoid submitting the CDK change when that happens.

skinny85 replied:
Is there a specific plan to enable the ACM option where it's currently not allowed?

Because if it's only a "maybe this will be allowed one day", I really don't like this sort of future-proofing.
And in that case, I would at least add runtime validation that checks whether `TlsValidationTrust.acm()` is passed somewhere it's not expected.

Seiya6329 replied:
Actually now that I think of this, I think users would get compile-time error if they attempt to pass `TlsValidationTrust.acm()` where it is not supporting `acm`.

For example, if a user attempt to pass `TlsValidationTrust` to `TlsListener`, we should expect the compile-time error because `ListenerTlsValidationContextTrust` does not support [`acm`](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-virtualnode-listenertlsvalidationcontexttrust.html).

skinny85 replied:
I'm not following...

Both `TlsValidationTrust.file()` and `TlsValidationTrust.acm()` have the same return type (`TlsValidationTrust`).

How is it possible that you can pass `TlsValidationTrust.file()` somewhere, but passing `TlsValidationTrust.acm()` in that same place would not compile?

Seiya6329 replied:
You are right on this. I went ahead and implemented the change to try inserting `TlsValidationTrust.acm()` but I did not get both compile-time and run-time error. 

As your suggested earlier, I will add run-time to check.

skinny85 replied:
Sounds good!

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts:
```diff
@@ -0,0 +1,185 @@
+import * as acmpca from '@aws-cdk/aws-acmpca';
+import { CfnVirtualGateway, CfnVirtualNode } from './appmesh.generated';
+
```
Thanks for sharing the suggestion. For this change, I will create two files by separating `TlsClientPolicy` and `TlsValidation` as advised.

Seiya6329 commented:
### REV4:
- Updating `README.md` file to fix the property names
- Reverting back the `client-policy.ts`

### DESIGN NOTES:
- Separating `TlsClientPolicy` and `TlsValidation` into their own files.

### Test:
`yarn run build` and `yarn run test` successful.

### BREAKING CHANGE: 
* **virtual-gateway:** major update on `ClientPolicy` object causes to break
* **virtual-node:** major update on `ClientPolicy` object causes to break

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: ea12b35d4ffc763b84cc15044905856284d319bb
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1612c9e2-a52c-4dd2-a8c2-191576f12d13%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 6a1185ad0dab4c4170e73d48c45af5b0aaa89ac3
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1d6372d5-6b54-47cb-845b-c716f7eb53ff%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

Seiya6329 commented on packages/@aws-cdk/aws-appmesh/lib/index.ts:
```diff
@@ -15,6 +15,7 @@ export * from './virtual-gateway';
 export * from './health-checks';
 export * from './tls-listener';
+export * from './tls-validation';
+export { TlsClientPolicy } from './client-policy';
```
@skinny85 - I am actually curious if this is better statement than saying `export * from` when the module is only exporting one item.

skinny85 replied:
No. Because then you have to consult a separate file to know whether a given exported member is part of the API of the module, or not.

This way, all that information is kept local to the file the class/interface is defined in.

skinny85 replied:
Actually, I believe the build might be failing because of this (it might be a JSII quirk that requires `export *`).

Seiya6329 commented:
### REV5:
- Fixing the merge conflict.

### Test:
`yarn run build` and `yarn run test` successful.

### BREAKING CHANGE: 
* **virtual-gateway:** major update on `ClientPolicy` object causes to break
* **virtual-node:** major update on `ClientPolicy` object causes to break

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 90e2101e4f223378b27dd52ba7d0a1003d685b5e
* Result: FAILED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e423ba21-5952-4ed5-b9f6-7b2a901c8d8e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/client-policy.ts:
```diff
@@ -1,113 +1,26 @@
-  FILE = 'file',
-}
+import { TlsValidation } from './tls-validation';
 
```
This file should be called `tls-client-policy.ts` (same way it's `tsl-validation.ts`, not `validation.ts`).

Seiya6329 replied:
Changed the name to `tls-client-policy.ts`

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/tls-validation.ts:
```diff
@@ -0,0 +1,136 @@
+   * Returns Trust context based on trust type.
+   */
+  public abstract bind(scope: Construct): TlsValidationTrustConfig;
+
```
No need for this extra empty line.

Seiya6329 replied:
Well noted. Thanks!

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts:
```diff
@@ -187,6 +188,7 @@ export class VirtualGateway extends VirtualGatewayBase {
     }
 
     const accessLogging = props.accessLog?.bind(this);
+    const backendDefaultsTlsClientPolicy = props.backendDefaults?.tlsClientPolicy;
```
Since this variable is used exactly once, I don't think it's very valuable. I would inline it.

Seiya6329 replied:
Inlined `props.backendDefaults?.tlsClientPolicy`.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/virtual-node.ts:
```diff
@@ -189,6 +190,7 @@ export class VirtualNode extends VirtualNodeBase {
     props.listeners?.forEach(listener => this.addListener(listener));
     const accessLogging = props.accessLog?.bind(this);
     const serviceDiscovery = props.serviceDiscovery?.bind(this);
+    const backendDefaultsTlsClientPolicy = props.backendDefaults?.tlsClientPolicy;
```
Since this variable is used exactly once, I don't think it's very valuable. I would inline it.

Seiya6329 replied:
Inlined it.

skinny85 reviewed (changes requested):
Looks great @Seiya6329! Since the build is failing (probably because of the lack of the `*` export), I stuck a few last-minute comments.

Make also sure to edit the PR description ([this comment](https://github.com/aws/aws-cdk/pull/14782#issue-648064519)) to be more inline with https://github.com/aws/aws-cdk/pull/14856#issue-651712686 . Pay special attention to how the "Breaking changes" section is formatted, and also remember to write from the perspective of your customers - remember that what you write there will be displayed in the ChangeLog of the project.

Seiya6329 commented:
#### REV6:
- Renaming `client-policy.ts` to `tls-client-policy.ts`
- Changing the import statement in `index.ts`
- Addressing nits

#### NOTE:
- Updated [this comment](https://github.com/aws/aws-cdk/pull/14782#issue-648064519) to align its written pattern to [this comment](https://github.com/aws/aws-cdk/pull/14856#issue-651712686).

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: d249bcf776eeef050f14d864696793f2df5eaa1a
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a7f4a3ad-2e1d-4e87-9262-26ab70f69e96%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 reviewed (approved):
Looks great @Seiya6329!

Seiya6329 commented:
Thanks @skinny85 for all the advice and guidance!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 4ab720aed052dfb78daa01056bcad69cc0438149
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=274c4f1e-5c82-4359-b48c-ce1f80831ad7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
